pkg/cover/backend/elf.go

```diff
@@ -40,85 +253,130 @@ func elfReadSymbols(module *Module, info *symbolInfo) ([]*Symbol, error) {
 	if err != nil {
-		return nil, nil, err
-	}
-	text := file.Section(".text")
-	if text == nil {
-		return nil, nil, fmt.Errorf("no .text section in the object file")
+		return nil, err
 	}
-	kaslr := file.Section(".rela.text") != nil
-	debugInfo, err := file.DWARF()
+	allDwarfFunctions, err := elfReadSymbolsFromDwarf(debugInfo, compileUnits)
 	if err != nil {
-		if module.Name != "" {
-			log.Logf(0, "ignoring module %v without DEBUG_INFO", module.Name)
-			return nil, nil, nil
-		}
-		return nil, nil, fmt.Errorf("failed to parse DWARF: %v (set CONFIG_DEBUG_INFO=y on linux)", err)
-	}
-
-	var pcFix pcFixFn
-	if kaslr {
-		pcFix = func(r [2]uint64) ([2]uint64, bool) {
-			if r[0] >= r[1] || r[0] < text.Addr || r[1] > text.Addr+text.Size {
-				// Linux kernel binaries with CONFIG_RANDOMIZE_BASE=y are strange.
-				// .text starts at 0xffffffff81000000 and symbols point there as
-				// well, but PC ranges point to addresses around 0.
-				// So try to add text offset and retry the check.
-				// It's unclear if we also need some offset on top of text.Addr,
-				// it gives approximately correct addresses, but not necessary
-				// precisely correct addresses.
-				r[0] += text.Addr
-				r[1] += text.Addr
-				if r[0] >= r[1] || r[0] < text.Addr || r[1] > text.Addr+text.Size {
-					return r, true
+		return nil, err
+	}
+	var symbols []*Symbol
+	for _, ds := range allDwarfFunctions {
+		if len(ds.Ranges) == 0 {
+			continue
+		}
+		start := ds.Ranges[0][0]
+		end := ds.Ranges[len(ds.Ranges)-1][1]
+		var ranges [][2]uint64
+		for _, r := range ds.Ranges {
+			ranges = append(ranges, [2]uint64{
+				r[0] + module.Addr,
+				r[1] + module.Addr,
+			})
+		}
+		flag := start >= text.Addr && end <= text.Addr+text.Size
+		if flag {
+			start1 := start + module.Addr
+			end1 := end + module.Addr
+			symbols = append(symbols, &Symbol{
+				Module: module,
+				ObjectUnit: ObjectUnit{
+					Name: ds.Name,
+				},
+				Start:  start1,
+				End:    end1,
+				Ranges: ranges,
+				Inline: ds.Inline,
+				Unit: &CompileUnit{
+					Module: module,
+					ObjectUnit: ObjectUnit{
+						Name: ds.DeclFile,
+					},
+				},
+			})
+		}
+	}
+	return symbols, nil
+}
+
+func elfReadSymbolsFromDwarf(debugInfo *dwarf.Data, compileUnits []*DwarfCompileUnit) ([]*DwarfFunction, error) {
+	var allSubprograms []*DwarfFunction
+	var allInlineSubroutines []*DwarfFunction
+	for _, cu := range compileUnits {
+		errc := make(chan error, 1)
+		go func(cu *DwarfCompileUnit) {
+			subprograms, err := readAllSubprograms(debugInfo, cu)
+			if err != nil {
+				errc <- err
+				return
+			}
+			allSubprograms = append(allSubprograms, subprograms...)
+			inlineSubprograms := make(map[dwarf.Offset]*DwarfFunction)
+			for _, sp := range subprograms {
+				if sp.Inline {
+					inlineSubprograms[sp.Offset] = sp
+				}
+			}
+			for _, sp := range subprograms {
+				if sp.Inline {
+					continue
+				}
+				erric := make(chan error, 1)
+				go func(sp *DwarfFunction) {
```

Comment:
It's necessary. From the test without gorountiue inside goroutine, it takes more than 30minutes to decode Android kernel + modules. But with it, it decreased to 1.5 minutes which I think it can be acceptable. (Tested on nproc==4 machine)